### DISCUSSION TITLE
Error and warning devtools::install

### DISCUSSION THREAD

**Discussion by shayavh** [IBAHCM/RPiR#343] [Archive] [answered]
Hi @richardreeve @soniamitchell,

I tried doing `devtools::install()` and when it asks for updates I try all and it then gives me an error:
`Error in download.file(url, destfile, method, mode = "wb", ...) : 
  (converted from warning) cannot open URL 'https://cran.rstudio.com/bin/macosx/contrib/4.0/tibble_3.1.0.tgz': HTTP status was '404 Not Found'
Error in download.packages(pkgs, destdir = tmpd, available = available,  : 
  (converted from warning) download of package ‘tibble’ failed`

Besides when I try doing it without the updates it gives me this warning
`Warning messages:
1: replacing previous import ‘dplyr::filter’ by ‘stats::filter’ when loading ‘shayavhDiversity’ 
2: replacing previous import ‘dplyr::lag’ by ‘stats::lag’ when loading ‘shayavhDiversity’ `

I think this warning is there because `dplyr` masks `filter` and `lag` from `stats`, but I have no idea how to get rid of it or if it even matters. 
Could you help me out here?

Thanks!

**Comment by richardreeve** [comment]
Well, that's very weird. You're right that that file seems to be missing from CRAN. I would just not select any updates when you do `devtools::install()`. You don't need to do any. Our last updates to RPiR were weeks ago now.

Your warning is interesting. It has spotted something that might cause problems in the real world. I don't know whether it does for you though. In your code, you are presumably using some functions from the stats and dplyr packages? (possibly your BCI package uses dplyr). If so do you know exactly what?

If none of your code uses `filter()`, then it doesn't matter and you could choose to ignore it.

However, if you do use `filter()` or you want to get rid of the warning because it's inelegant(!), you can do it like we do in RPiR here:

https://github.com/IBAHCM/RPiR/blob/5700c1e51ac26bb279d7e4bbf50ab61236eeed1a/R/RPiR-package.R#L66

Instead of writing:
```{R}
#' @import stats
```
in the package documentation file (or wherever you have put the import statement), you can individually import the exact function that you use (here `statfun()`):

```{R}
#' @importFrom stats statfun
```

Does that make sense? This is the technically correct thing to do for purists because it allows you to say not that you'll use a whole package, but you'll use a specific function in a package.

**Reply by richardreeve** [reply]
PS The actual warning is essentially the opposite of this standard dplyr warning:
```{R}
> library(dplyr)

Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag
```

**Comment by soniamitchell** [comment, marked as answer]
Hey @shayavh, I'm not completely sure what's happening here. But, I removed `stats` as a dependency (since I have an older version on my machine) as it's a base R package and doesn't *really* need to be a dependency and that seems to have fixed it. Does it fix it on your machine?

**Reply by richardreeve** [reply]
Hah. True. Or if you don't use stats you could certainly do that. Sonia's right that you don't **need** to add stats as a dependency anyway because we can guarantee that it'll be installed (because it's a base package). However, officially you are supposed to. It's a bit of a fine distinction...

**Reply by soniamitchell** [reply]
Yea.. the CMD check asks for it.. but you’re not supposed to add dependencies on base packages. It’s a bit of a contradiction. Specifying the latest version is a bit strict either way. The tibble error caught me out though.. maybe they fixed it and I’m too late to see it? Otherwise I’ve no idea why removing stats would fix it...

**Reply by shayavh** [reply]
Thanks @soniamitchell @richardreeve, I have now removed `stats` from all my functions and demos and removed them as dependencies. It works now! :)
